.github/generate-doc.py

```diff
@@ -12,6 +14,137 @@ class Colors:
     RESET = '\033[0m'
 
 
+def format_markdown_file(filename):
+    with open(filename, 'r+') as file:
+        content = file.read()
+        formatted_content = markdown2.markdown(content)
+    with open(filename, 'w') as file:
+        file.write(formatted_content)
+    replace_string_in_markdown(filename, "<p>", "")
+    replace_string_in_markdown(filename, "</p>", "")
+    replace_string_in_markdown(filename, "<em>", "_")
+    replace_string_in_markdown(filename, "</em>", "_")
+
+
+def contains_subsection(file_path, target_subsection_title):
+    section_pattern = re.compile(r'^\s*#{1}\s+(.*)$', re.MULTILINE)
+    subsection_pattern = re.compile(r'^\s*#{2}\s+(.*)$', re.MULTILINE)
+
+    with open(file_path, 'r', encoding='utf-8') as file:
+        markdown_text = file.read()
+
+    section_matches = section_pattern.finditer(markdown_text)
+
+    for match in section_matches:
+        section_title = match.group(1)
+        section_start = match.end()
+
+        if section_title.lower() == target_subsection_title.lower():
+            return bool(subsection_pattern.search(markdown_text, section_start))
+
+    return False
+
+
+def replace_string_in_markdown(file_path, old_string, new_string):
+    with open(file_path, 'r', encoding='utf-8') as file:
+        content = file.read()
+
+    modified_content = content.replace(old_string, new_string)
+
+    with open(file_path, 'w', encoding='utf-8') as file:
+        file.write(modified_content)
```

Comment:
I think you could save on file openings and lines of code here, because writing also should allow reading the file:
```suggestion
    with open(file_path, 'w', encoding='utf-8') as file:
        content = file.read()
        file.write(content.replace(old_string, new_string))
```

Reply:
`io.UnsupportedOperation: not readable`
it does not seem to work